pkg/services/ngalert/migration/service.go

```diff
@@ -49,9 +69,464 @@ func ProvideService(
+		return &summary, nil
+	})
+}
+
+// GetOrgMigrationState returns the current migration state for an org. This is a potentially expensive operation as it
+// requires re-hydrating the entire migration state from the database against all current alerting resources.
+func (ms *migrationService) GetOrgMigrationState(ctx context.Context, orgID int64) (*definitions.OrgMigrationState, error) {
+	if migrated, err := ms.migrationStore.IsMigrated(ctx, orgID); err != nil || !migrated {
+		return &definitions.OrgMigrationState{OrgID: orgID}, err
+	}
+	var resState *definitions.OrgMigrationState
+	err := ms.store.InTransaction(ctx, func(ctx context.Context) error {
+		var err error
+		dState, err := ms.migrationStore.GetOrgMigrationState(ctx, orgID)
+		if err != nil {
+			return err
+		}
+
+		cfg, err := ms.migrationStore.GetAlertmanagerConfig(ctx, orgID)
+		if err != nil {
+			return fmt.Errorf("get alertmanager config: %w", err)
+		}
+		amConfig := migmodels.FromPostableUserConfig(cfg)
+
+		// Hydrate the slim database model.
+		resState = &definitions.OrgMigrationState{
+			OrgID:              dState.OrgID,
+			MigratedDashboards: make([]*definitions.DashboardUpgrade, 0, len(dState.MigratedDashboards)),
+			MigratedChannels:   make([]*definitions.ContactPair, 0, len(dState.MigratedChannels)),
+		}
+
+		channels, err := ms.migrationStore.GetNotificationChannels(ctx, orgID)
+		if err != nil {
+			return fmt.Errorf("get notification channels: %w", err)
+		}
+
+		existingChannels := make(map[int64]struct{})
+		for _, channel := range channels {
+			existingChannels[channel.ID] = struct{}{}
+			pair := &definitions.ContactPair{
+				LegacyChannel: fromAlertNotification(channel),
+				Error:         "channel not upgraded",
+			}
+
+			if p, ok := dState.MigratedChannels[channel.ID]; ok {
+				if p.NewReceiverUID != "" {
+					if recv, ok := amConfig.GetReceiver(p.NewReceiverUID); ok {
+						if route, ok := amConfig.GetLegacyRoute(recv.Name); ok {
+							pair.ContactPointUpgrade = fromContactPointUpgrade(recv, route)
+						}
+					}
+				}
+				pair.Error = p.Error
+			}
+
+			resState.MigratedChannels = append(resState.MigratedChannels, pair)
+		}
+
+		// Now we check the inverse, for channels that were migrated but no longer exist.
+		for _, p := range dState.MigratedChannels {
+			if _, ok := existingChannels[p.LegacyID]; !ok {
+				pair := &definitions.ContactPair{
+					LegacyChannel: &definitions.LegacyChannel{ID: p.LegacyID},
+					Error:         "channel no longer exists",
+				}
+				if p.NewReceiverUID != "" {
+					if recv, ok := amConfig.GetReceiver(p.NewReceiverUID); ok {
+						if route, ok := amConfig.GetLegacyRoute(recv.Name); ok {
+							pair.ContactPointUpgrade = fromContactPointUpgrade(recv, route)
+						}
+					}
+				}
+				resState.MigratedChannels = append(resState.MigratedChannels, pair)
+			}
+		}
+
+		// We preload information in bulk for performance reasons.
+		mappedRules, err := ms.migrationStore.GetSlimAlertRules(ctx, orgID)
```

Comment:
I think legacyAlertRules could be a better name. Not sure I understand `mapped` prefix.